### PR TITLE
Support initialViewState updates

### PR DESCRIPTION
**Pull request by Pessimistress**
#### Background

We currently have two models of view state management:

- Stateless:

  ```jsx
  <DeckGL controller={true}
      viewState={this.state.viewState}
      onViewStateChange={({viewState}) => this.setState({viewState})} />
  ```

- Stateful

  ```jsx
  <DeckGL controller={true}
      internalViewState={INITIAL_VIEWSTATE} />
  ```

Currently, if the user wants to manipulate the view state outside of the controller at any time (e.g. click a button and fly to another city), they have to use the stateless model.

The main concern is that this breaks the reactive paradigm (potentially modifying props outside of the React render cycle). Maybe we don't expose this method on the React component?

Related: [Imperative API RFC](https://github.com/uber/deck.gl/blob/master/dev-docs/RFCs/v7.x/imperative-api-rfc.md)

#### Change List
- Add `deck.setViewState` and `DeckGL.setViewState`
- Docs
- Update tile-3d and playground examples to use the new API

### CODE REVIEW

File: modules/react/src/deckgl.js [hunk 1]
```diff
@@ -73,6 +71,11 @@ export default class DeckGL extends React.Component {
         })
       );
     this._updateFromProps(this.props);
+
+    // Bind public methods
```
Comment (ibgreen) [inline modules/react/src/deckgl.js]: Nits:
1. This change seems unrelated.
2. This is a little too terse for my tastes, I would just keep and bind the methods here. deck is complicated and as a rule I think it is worth going for simpler standardized code.

**Comment by Pessimistress** [comment]
> What is the main advantage of a separate method as compared to just refining the semantics of `deck.setProps({viewState})`?

As of v8, `initialViewState` has no effect if `viewState` is specified. The concept of `props` is originated from the reactive model, whereas the rendered view state should never deviate from the user props if provided. Allowing internally tracked view state to overwrite `props.viewState` was the source of many confusing behavior in v7, especially React (where we call `deck.setProps` with not just the specified props, but merged with default props at every render).

**Comment by ibgreen** [comment]
As of v8, initialViewState has no effect if viewState is specified.

Yes, sounds like a good change, keeping this simple is highly desirable!

That said, it is hard to "strongly endorse" this new function, probably mainly because it is not part of any API "pattern", it is just a one-off solution, and sets a possibly strange precedent for our API.

Looking for alternatives (to update the auto tracked view state without kicking out of tracking mode) the main thing I can think of:

- If not already done, we could define `deck.setProps({initialViewState: newViewState})` to reset the tracked `viewState`, if new `initialViewState` is different from previous `initialViewState`? 

But this would be a partial replacement only, as it would not handle the case where the user wants to reset the view state to the current initial view state.

Not sure I have more input to give, do what you think is best.

**Comment by ibgreen** [comment]
> Not sure I have more input to give...

Of course as soon as I wrote that I realized:

- For React and JS deck.gl programmers, `initialViewState` is a nice convenience.
- However, the primary use case for `initialViewState` is really declarative usage in `deck.gl JSON` and `pydeck`, etc as these bindings do not (currently) have any way to define useful `onViewStateChange` callbacks.
- These language bindings will not easily get the benefit of a new `setViewState` method. They are dependent on the prop updates mechanism.

This should encourage us to think extra hard about how to design this in a declarative way, since we would want e.g. pydeck applications to be able to move the view state around in response to other interactions, data loads etc.

Thoughts?

**Comment by Pessimistress** [comment]
I think watching for `initialViewState` change might work. I'll give it a go.

File: docs/api-reference/deck.md [hunk 2]
```diff
@@ -121,11 +121,14 @@ Transitions between two viewState objects can also be achieved by providing set
...
 
+If the `initialViewState` prop changes, the internally tracked view state will be updated to match the new "initial" view state.
+
 Notes:
 
 * The `props.onViewStateChange` callback will still be called, if provided.
 * If `props.viewState` is supplied by the application, the supplied `viewState` will always be used, "shadowing" the `Deck` component's internal `viewState`.
-* In simple applications, use of the `initialViewState` prop can avoid the need track the view state in the application .
+* In simple applications, use of the `initialViewState` prop can avoid the need track the view state in the application.
+* Any new `initialViewState` value is compared shallowly with the previous value. Therefore, it is especially important for React applications to use a constant or component state as the `initialViewState` value, in order to avoid any unexpected "reset" of the camera.
```
Comment (ibgreen) [inline docs/api-reference/deck.md]: I think the shallow compare should be discussed. For the pydeck/json case, where the app builds new JSON payloads to control the visualization, the shallow equality will always be `false`.

I think the best compromise might be to compare `initialViewState` (if supplied) deeply.

File: modules/core/src/lib/deck.js [hunk 3]
```diff
@@ -241,6 +242,7 @@ export default class Deck {
     }
   }
 
+  /* eslint-disable complexity */
```
Comment (ibgreen) [inline modules/core/src/lib/deck.js]: Nit: Use `// eslint-disable-next-line complexity`

File: modules/core/src/lib/deck.js [hunk 4]
```diff
@@ -298,6 +304,7 @@ export default class Deck {
 
     this.stats.get('setProps Time').timeEnd();
   }
+  /* eslint-enable complexity */
```
Comment (ibgreen) [inline modules/core/src/lib/deck.js]: ...then this can be dropped

**Comment by jonenst** [comment]
After changing the initialViewState prop, is it supposed to call onViewStateChange callback? In my tests it doesn't call it but I was expecting it to be called.
